ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +203,429 @@ let order_f (_, vm_rec) =
+  | Some (0, host_size, host) -> (
+    match (vm_size <= host_size, vm_can_boot_on_host vm host) with
+    | true, true ->
+        Some host
+    | _, _ ->
+        select_host_for_no_breach_plan vm vm_size vm_can_boot_on_host
+          (HostsSet.remove (0, host_size, host) no_resident_hosts_set)
+  )
+  | _ ->
+      None
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context vm_can_boot_on_host
+    anti_affinity_vms pool_state mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  match anti_affinity_vms with
+  | [] ->
+      mapping
+  | (vm, vm_size, group) :: remaining_vms -> (
+      debug
+        "Spread evenly plan: try to plan for the anti-affinity VM (%s %s %s)."
+        (Ref.string_of vm)
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (Db.VM_group.get_name_label ~__context ~self:group) ;
+      match
+        anti_affinity_hosts_set group pool_state
+        |> select_host_for_spread_evenly_plan vm vm_size vm_can_boot_on_host
+      with
+      | None ->
+          debug
+            "Spread evenly plan: no host can run this anti-affinity VM. Stop \
+             the planning as there won't be a valid plan for this VM." ;
+          raise (Api_errors.Server_error (Api_errors.no_hosts_available, []))
+      | Some h ->
+          debug
+            "Spread evenly plan: choose the host with the least free memory \
+             which can run the vm: (%s %s)."
+            (Ref.string_of h)
+            (Db.Host.get_name_label ~__context ~self:h) ;
+          vm_anti_affinity_spread_evenly_plan ~__context vm_can_boot_on_host
+            remaining_vms
+            (pool_state |> update_anti_affinity_pool_state vm_size group h)
+            ((vm, h) :: mapping)
+    )
+
+(** Try to get a no breach plan for anti-affinity VMs (for each anti-affinity group, there are at least
+    2 hosts having running VMs in the group): 
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, try to select a host for it so that there are at least 2 hosts which
+       has running VMs in the same anti-affinity group. If there are already 2 hosts having running VMs
+       in that group, skip planning for the VM. *)
+let rec vm_anti_affinity_no_breach_plan ~__context total_hosts
```

Comment:
There are some unnecessary parameters for this function. Say,
Initially, we have 
```
A: HostMap: host -> free_mem
B: GroupMap: group -> (HostMap : host -> vm_cnt)
```
The `plan_for_spead_evenly` requires:
```
A: HostMap: host -> free_mem
B: GroupMap: group -> (HostMap : host -> vm_cnt)
B -> C: GroupMap: group -> (vm_cnt, free_mem, h)
```

The `plan_for_no_breach` requires:
```
A: HostMap: host -> free_mem
B -> D: GroupMap: group -> (hosts, inhabited_hosts)
```
And the `B -> D` is as simple as:
```
  GroupMap.map (fun hosts ->
    let m1, m2 = HostMap.partition (fun _ vm_cnt -> vm_cnt = 0) hosts in
    (List.map fst HostMap.bindings, HostMap.cardinal m2)
  ) 
```

Reply:
To change it, I think the required info for `no_breach` would be like:
`
type anti_affinity_no_breach_pool_state = {
    host_size_map: int64 Xapi_vm_helpers.HostMap.t
  ; grp_no_resident_hosts_set: HostsSet.t VMGroupMap.t
  ; grp_resident_hosts_count_map: int VMGroupMap.t
}

`